Mathlib/Algebra/Module/Zlattice.lean

```diff
@@ -258,3 +297,143 @@ protected theorem isAddFundamentalDomain [Finite ι] [MeasurableSpace E] [OpensM
 end Real
 
 end Zspan
+
+section Zlattice
+
+open Submodule
+
+variable (K : Type _) [NormedLinearOrderedField K] [HasSolidNorm K] [FloorRing K]
+variable {E : Type _} [NormedAddCommGroup E] [NormedSpace K E] [FiniteDimensional K E]
+variable [ProperSpace E] {L : AddSubgroup E} [DiscreteTopology L]
+variable (hs : span K (L : Set E) = ⊤)
+
+theorem Zlattice.FG : AddSubgroup.FG L := by
+  suffices (AddSubgroup.toIntSubmodule L).FG by exact (fg_iff_add_subgroup_fg _).mp this
+  obtain ⟨s, ⟨h_incl, ⟨h_span, h_lind⟩⟩⟩ := exists_linearIndependent K (L : Set E)
+  -- Let `s` be a maximal `K`-linear independent family of elements of `L`. We show that
+  -- `L` is finitely generated (as a ℤ-module) because it fits in the exact sequence
+  -- `0 → span ℤ s → L → L / span ℤ s → 0` with `span ℤ s` and `L / span ℤ s` finitely generated.
+  refine fg_of_fg_map_of_fg_inf_ker (span ℤ s).mkQ ?_ ?_
+  · -- Let `b` be the `K`-basis of `E` formed by the vectors in `s`. The elements of
+    -- `L / span ℤ s = L /span ℤ b` are in bijection with elements of `L ∩ fundamentalDomain b`
+    -- so there are finitely many since `fundamentalDomain b` is bounded.
+    refine fg_def.mpr ⟨map (span ℤ s).mkQ (AddSubgroup.toIntSubmodule L), ?_, span_eq _⟩
+    let b := Basis.mk h_lind (by
+      rw [← hs, ← h_span]
+      exact span_mono (by simp only [Subtype.range_coe_subtype, Set.setOf_mem_eq, subset_rfl]))
+    rw [show span ℤ s = span ℤ (Set.range b) by simp [Basis.coe_mk, Subtype.range_coe_subtype]]
+    have : Fintype s := Set.Finite.fintype h_lind.finite
+    refine Set.Finite.of_finite_image (f := ((↑) : _ →  E) ∘ Zspan.QuotientEquiv b) ?_
+      (Function.Injective.injOn (Subtype.coe_injective.comp (Zspan.QuotientEquiv b).injective) _)
+    have : Set.Finite ((Zspan.fundamentalDomain b) ∩ L) :=
+      Metric.Finite_bounded_inter_isClosed (Zspan.fundamentalDomain_bounded b) inferInstance
+    refine Set.Finite.subset this ?_
+    rintro _ ⟨_, ⟨⟨x, ⟨h_mem, rfl⟩⟩, rfl⟩⟩
+    rw [Function.comp_apply, mkQ_apply, Zspan.quotientEquiv_apply_mk, Zspan.fract_restrict_apply]
+    refine ⟨?_, ?_⟩
+    · exact Zspan.fract_mem_fundamentalDomain b x
+    · rw [Zspan.fract, SetLike.mem_coe, sub_eq_add_neg]
+      refine AddSubgroup.add_mem _ h_mem
+        (neg_mem (Set.mem_of_subset_of_mem ?_ (Subtype.mem (Zspan.floor b x))))
+      rw [show (L : Set E) = AddSubgroup.toIntSubmodule L by rfl]
+      rw [SetLike.coe_subset_coe, Basis.coe_mk, Subtype.range_coe_subtype, Set.setOf_mem_eq]
+      exact span_le.mpr h_incl
+  · -- `span ℤ s` is finitely generated because `s` is finite
+    rw [ker_mkQ, inf_of_le_right (span_le.mpr h_incl)]
+    exact fg_span (LinearIndependent.finite h_lind)
+
+theorem Zlattice.module_finite : Module.Finite ℤ L :=
+  Module.Finite.iff_addGroup_fg.mpr ((AddGroup.fg_iff_addSubgroup_fg L).mpr (FG K hs))
+
+theorem Zlattice.module_free : Module.Free ℤ L := by
```

Comment:
These should be instances

Reply:
Yes, Lean does not like these much as instances :)